packages/components/src/components/SquigglePlayground.tsx

```diff
@@ -420,10 +304,137 @@ export const SquigglePlayground: FC<PlaygroundProps> = ({
+              playing && "animate-spin",
+              stale ? "text-indigo-500" : "text-gray-400"
+            )}
+          />
+        </button>
+      )}
+      <Toggle
+        texts={["Autoplay", "Paused"]}
+        icons={[CheckCircleIcon, PauseIcon]}
+        status={autoplay}
+        onChange={onAutoplayChange}
+      />
+    </div>
+  );
+};
+
+export const SquigglePlayground: FC<PlaygroundProps> = ({
+  initialSquiggleString = "",
+  height = 500,
+  showTypes = false,
+  showControls = false,
+  showSummary = false,
+  logX = false,
+  expY = false,
+  code: controlledCode,
+  onCodeChange,
+  onSettingsChange,
+  showEditor = true,
+}) => {
+  const [uncontrolledCode, setUncontrolledCode] = useState(
+    initialSquiggleString
+  );
+  const [renderedCode, setRenderedCode] = useState(""); // used only if autoplay is false
+  const code = controlledCode ?? uncontrolledCode;
+
+  const [imports, setImports] = useState({});
+
+  const {
+    register,
+    control,
+    setValue: setFormValue,
+  } = useForm({
+    resolver: yupResolver(schema),
+    defaultValues: {
+      sampleCount: 1000,
+      xyPointLength: 1000,
+      chartHeight: 150,
+      showTypes,
+      showControls,
+      logX,
+      expY,
+      showSummary,
+      showEditor,
+      leftSizePercent: 50,
+      showSettingsPage: false,
+      diagramStart: 0,
+      diagramStop: 10,
+      diagramCount: 20,
+      autoplay: true,
+    },
+  });
+  const vars = useWatch({
+    control,
+  });
+
+  useEffect(() => {
+    onSettingsChange?.(vars);
+  }, [vars, onSettingsChange]);
+
+  const chartSettings = useMemo(
+    () => ({
+      start: Number(vars.diagramStart),
+      stop: Number(vars.diagramStop),
+      count: Number(vars.diagramCount),
+    }),
+    [vars.diagramStart, vars.diagramStop, vars.diagramCount]
+  );
+
+  const env: environment = useMemo(
+    () => ({
+      sampleCount: Number(vars.sampleCount),
+      xyPointLength: Number(vars.xyPointLength),
+    }),
+    [vars.sampleCount, vars.xyPointLength]
+  );
+
+  const [playing, setPlaying] = useState(false); // used in manual play mode only
+
+  // This part is tricky and fragile; we need to re-render first to make sure that the icon is spinning,
+  // and only then evaluate the squiggle code (which freezes the UI).
+  // Also note that `useEffect` execution order matters here.
+  // Hopefully it'll all go away after we make squiggle code evaluation async.
+  useEffect(() => {
+    if (renderedCode === code && playing) {
+      // It's not possible to put this after `setRenderedCode(code)` below because React would apply
+      // `setPlaying` and `setRenderedCode` together and spinning icon will disappear immediately.
+      setPlaying(false);
+    }
+  }, [renderedCode, code, playing]);
+
+  useEffect(() => {
+    if (!vars.autoplay && playing) {
+      setRenderedCode(code); // TODO - force play even if code hasn't changed
+    }
+  }, [vars.autoplay, code, playing]);
+
+  const play = () => {
```

Comment:
small note; but do you need the {} braces here?

Reply:
> Small note; but do you need the {} braces here?

`setX` returns void, so it shouldn't matter in this case, but usually I don't skip `{}` if function's body is a single call, to make sure that it doesn't return anything. (E.g., sometimes the function gets passed to a component which expects `() => void`, and passing `() => SomethingElse` would fail type checks).

Reply:
> Also, go for a more descriptive name, like, startPlaying. This isn't running once, it's just changing the state once.

This seems like an implementation detail to me? `play` is the function that should be called to trigger the run, and it effectively does that (by causing `useEffect`s to fire), and places in the code where it's used shouldn't care how the run is triggered.

I know that it doesn't cause the re-run if code hasn't changed, but that's in TODO and I hope to fix it soon.